### PR TITLE
First draft of the database logic

### LINKED ISSUE #410: Draft a skeleton / first version of the DB population process
## 🚀 Feature

Start minimally, borrow ideas form alredy existing database processes / entrypoint.

Already, we can see at least two different subtasks
- Initialise the DB
- Incremental population step

But there's a lot more to think about...
- different tasks / subtasks
- how to organise them?
- are we going to write entrypoints?

=> Brainstorm / make a minimal proposal then discuss.

## Motivation

Please outline the motivation for the proposal. Is your feature request related to a problem? e.g., I'm always frustrated when [...]. If this is related to another GitHub issue, please link here too.

## Pitch

A clear description of what you want to happen.

## Alternatives

Think about any alternative solutions or features that could be used.
Then, write here a clear list of alternatives.

## Additional context
Add any other screenshot, context, or information about the problem here.

### PR DESCRIPTION
**Pull request by jankrepl**
Closes #410 

## Description (updated 11.8.2021)
***What was done***
1. Implement entrypoint `bbs_database` with nested commands (using `argparse` for consistency)
2. Implement database creation  - `bbs_database init`
    - Single table `articles` with a single column `title`
3. Implement parsing of a single article - `bbs_database parse`
4. Implement adding of a single article to a database `bbs_database add`
5. Write tests 

***What is missing***
1. Decide on and implement an actual database schema
2. MySQL support
3. `parse` and `add` only work on a single article at a time 
4. `add` does no input validation (duplicats/unique ID)
4. Add more commands (e.g. `backup`,  `add-many`, `delete`, ...)

## How to test?
I wrote tests so one way to test this is to run  `pytest tests/test_entrypoint/database/` (however, the CI already does this).

To get familiar with the new entrypoint you can run something along these lines
```bash
# Install `bluesearch` since there is a new entrypoint `bbs_database`
pip install -e .

# Get familiar with the CLI
bbs_database -h 

# Initialize an empty database (with the correct schema)
DB=temp.db
bbs_database init $DB

# Parse a couple of articles
PARSER=CORD19ArticleParser 
bbs_database parse $PARSER tests/data/cord19_v35/document_parses/pmc_json/PMC7140272.xml.json parsed_1.pkl
bbs_database parse $PARSER tests/data/cord19_v35/document_parses/pmc_json/PMC7186928.xml.json parsed_2.pkl
bbs_database parse $PARSER tests/data/cord19_v35/document_parses/pmc_json/PMC7223769.xml.json parsed_3.pkl

# Add to the database
bbs_database add $DB parsed_1.pkl
bbs_database add $DB parsed_2.pkl
bbs_database add $DB parsed_3.pkl

# Verify it worked
sqlite3 $DB

```
And then inside of the `sqlite3` console
```
sqlite> .tables
sqlite> SELECT * FROM articles;
```

## Comments
I made multiple design decisions (especially when it comes to structuring the CLI). Feel free to share your opinion and if necessary we can completely change it:)

## Checklist

- [x] This PR refers to an issue from the [issue tracker](https://github.com/BlueBrain/Search/issues).
  (if it is not the case, please create an issue first).
- [x] Unit tests added.
  (if needed)
- [ ] Documentation and `whatsnew.rst` updated.
  (if needed)
- [x] `setup.py` and `requirements.txt` updated with new dependencies.
  (if needed)
- [x] Type annotations added.
  (if a function is added or modified)
- [x] All CI tests pass.

### CODE REVIEW

File: src/bluesearch/entrypoint/database/add.py
Comment (Stannislav) [inline src/bluesearch/entrypoint/database/add.py:69]: Just a small caveat here: if a subclass is in a different module, then this module needs to be imported first, otherwise the subclass won't be listed. At the moment we have all subclasses in the same module, but it's possible we'll refactor in the future, and it's something easy to miss.

Without trying to start an argument, just wanted to mention again that I'm not a super big fan of dynamic name resolutions and turning strings into python code. So everytime I see `__import__`, `getattr`, `eval`, `vars(...)`, `func(**kwargs)`, etc. a part of me gets a bit sad. I think this makes the code less readable and harder to reason about, probably also stops IDEs/mypy from properly resolving objects too. It's a bit personal though, so no offence!

I just feel that if there are O(1) possible options like in the present case, then they can be easily enumerated/listed. Personally I would probably try and use dictionaries / enums / some kind of simple registry.

Anyway, no offence, just contributing my 2 cents of philosophical thought, we can carry on 😄

File: src/bluesearch/entrypoint/database/add.py
Comment (Stannislav) [inline src/bluesearch/entrypoint/database/add.py:85]: What was the difference between `engine.connect()` and `engine.begin()` again? I always forget 🙈

File: tests/test_entrypoint/database/test_add.py
Comment (Stannislav) [inline tests/test_entrypoint/database/test_add.py:44]: Isn't this equal to the `jsons_path` fixture in the parameters, that's actually unused at the moment?

**Review by Stannislav — COMMENTED** [review]
Hey @jankrepl thanks a lot for this darft implementation. I think once it's merged it'll really allow us to iterate on it and to add new features :)

**Review by EmilieDel — COMMENTED** [review]
Hello @jankrepl, 

Thanks a lot for doing this PR! Maybe two global questions/thoughts: 
- How do you see the pipeline if we want to include a new batch of articles in the database ? Is it something we will handle at the command line level ? Or should we consider  handling also this case in the code itself ?
- Should the file format and data structure be something that user needs to specify himself (currently through the parser) or should we handle this automatically in the code ? In a way, the extension of the file should already give information about the way we need to load the file for example. Moreover, if we ask the user to specify the parser, what should we do if the coupling seems wrong ?  

What is your opinion about this ?

File: setup.py
Comment (FrancescoCasalegno) [reply setup.py:126]: > I guess this choice was just to be in line with the other entrypoint names, so fair enough.

Not against using `-` instead of `_`, but I would be for doing that in a separate PR, and keep it consistent as Jan did here for the moment.

Comment (Stannislav) [reply setup.py:126]: yep, agree about separate PR

File: src/bluesearch/entrypoint/database/add.py
Comment (Stannislav) [inline src/bluesearch/entrypoint/database/add.py:78]: Isn't "r" already the default anyway?

**Review by pafonta — COMMENTED** [review]
Thanks for the proposition on how to add articles to the SQL database.

From my experience, coupling the parsing and the loading in the database is not a best practice (i.e. `bbs_database add <database> <parser class> <data path>`). So, I would be against it. But lets see what the team thinks.

Besides, I have the same concerns as here https://github.com/BlueBrain/Search/pull/412#discussion_r681066565 and the questions as here https://github.com/BlueBrain/Search/pull/412#pullrequestreview-720920681.

File: src/bluesearch/entrypoint/database/parent.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/parent.py:12]: Good point. Just to be sure, are you using `typeshed` as the reference?
https://github.com/python/typeshed/blob/76eeeee4c4375b32db9a6d2cc3f5a57d756ab1d6/stdlib/argparse.pyi#L145

I am going to change it.

Comment (jankrepl) [reply src/bluesearch/entrypoint/database/parent.py]: Done in 1261dee8696ec37acf3b83a04da25823ba7ad15c

**Comment by jankrepl** [comment]
> Thanks for the proposition on how to add articles to the SQL database.
> 
> From my experience, coupling the parsing and the loading in the database is not a best practice (i.e. `bbs_database add <database> <parser class> <data path>`). So, I would be against it. But lets see what the team thinks.
> 
> Besides, I have the same concerns as here [#412 (comment)](https://github.com/BlueBrain/Search/pull/412#discussion_r681066565) and the questions as here [#412 (review)](https://github.com/BlueBrain/Search/pull/412#pullrequestreview-720920681).

Thanks for the review:) Could you maybe provide a more constructive feedback? What is the best practice and ideally could you link some open-source projects that implement it? I would be more than happy to make major changes:)

File: src/bluesearch/entrypoint/database/parent.py
Comment (Stannislav) [reply src/bluesearch/entrypoint/database/parent.py]: Yes, I think PyCharm uses `typeshed` for it's type hints.

Thanks for fixing :)

**Comment by jankrepl** [comment]
Thank you for the review!

> * How do you see the pipeline if we want to include a new batch of articles in the database ? Is it something we will handle at the command line level ? Or should we consider  handling also this case in the code itself ?

> * Should we add a duplicate check at some point in the pipeline ? I know we have an opened issue about generating unique IDs for the database data.

I actually quickly mentioned these in the description. I intentionally did not implement anything and just went for the simplest solution. IMO both of these require design choices that we should decide on together (and probably in a different PR).

> * Should the file format and data structure be something that user needs to specify himself (currently through the parser) or should we handle this automatically in the code ? In a way, the extension of the file should already give information about the way we need to load the file for example. Moreover, if we ask the user to specify the parser, what should we do if the coupling seems wrong ?

Interesting idea! What about cases where we have have multiple parsers for the same extension though? I am not sure if that would make sense. Anyway, I guess the current setup is more flexible and verbose at the same time. We can definitely change it in the future. And regarding sanity checks, I would argue this should happen inside of the `ArticleParser` subclasses.

File: src/bluesearch/entrypoint/database/parent.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/parent.py]: The `_` was done here f72fabc68ef31116bc1119386f5525e0098db113

Comment (jankrepl) [reply src/bluesearch/entrypoint/database/parent.py]: @Stannislav I agree that `argparse` is really awkward when it comes to creating subparsers. Thank you for the suggestion, I did not know one can do it that way. The approach I proposed was actually a result of me stumbling upon this: https://stackoverflow.com/questions/14988397/is-there-a-way-to-add-an-already-created-parser-as-a-subparser-in-argparse . Let me give this a second thought ....

File: src/bluesearch/entrypoint/database/add.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/add.py:85]: I think it has to do something with potential rollbacks in case an exception appears. https://docs.sqlalchemy.org/en/13/core/connections.html#using-transactions

Anyway, honestly, I feel like I use them interchangeably. Did we decide to stick to one of them?

File: tests/test_entrypoint/database/test_parent.py
Comment (jankrepl) [reply tests/test_entrypoint/database/test_parent.py:8]: For some reason this actually does not work.
```
parent.main([command, "--help"])
```
It raises SystemExit error when launched via `pytest`. However, works completely fine when run with plain `python`.

File: src/bluesearch/entrypoint/database/add.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/add.py]: Fixed in beeaf5511df03f60f424925f8422b3def13b3d97

File: src/bluesearch/entrypoint/database/init.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/init.py]: Fixed in beeaf5511df03f60f424925f8422b3def13b3d97

File: src/bluesearch/entrypoint/database/init.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/init.py:45]: I went for `sqlite` because this is just a draft and I did not want to spend too much time setting up `mysql` and tests

File: src/bluesearch/entrypoint/database/add.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/add.py]: Done in b8ff73b532140c2f2ad757693b1fe88c7df62abf

File: src/bluesearch/entrypoint/database/add.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/add.py:70]: I am totally fine with having an if statement:)

File: src/bluesearch/entrypoint/database/init.py
Comment (EmilieDel) [reply src/bluesearch/entrypoint/database/init.py:45]: Fair enough :)

File: tests/test_entrypoint/database/test_parent.py
Comment (Stannislav) [reply tests/test_entrypoint/database/test_parent.py:8]: Isn't that expected?

**Comment by pafonta** [comment]
> > Thanks for the proposition on how to add articles to the SQL database.
> > From my experience, coupling the parsing and the loading in the database is not a best practice (i.e. `bbs_database add <database> <parser class> <data path>`). So, I would be against it. But lets see what the team thinks.
> > Besides, I have the same concerns as here [#412 (comment)](https://github.com/BlueBrain/Search/pull/412#discussion_r681066565) and the questions as here [#412 (review)](https://github.com/BlueBrain/Search/pull/412#pullrequestreview-720920681).
> 
> Thanks for the review:) Could you maybe provide a more constructive feedback? What is the best practice and ideally could you link some open-source projects that implement it? I would be more than happy to make major changes:)

First of all, I am not asking for changes. My review has been requested for the current approach. Not decoupling parsing and loading is a common pitfall I have seen during my 8 years of professional experience working with data. My suggestion is then only to proactively prevent this common issue to happen in this project. But if the team chooses to move on, let's move on!

---

> Could you maybe provide a more constructive feedback?

The extensive example about decoupling parsing and loading is at the end of https://github.com/BlueBrain/Search/pull/383#pullrequestreview-687607172. It was part of the review of the PR introducing the `ArticleParser`.  The specific part of the example I am thinking about is:

```python
# Parsing and Loading articles is decoupled.

# Parse articles.
articles = parse_cord19(dirpath, nlp)

# Load articles in the database.
engine = create_engine("mysql+pymysql://<user>:<password>@<host>:<port>/<database>")
write_sql(articles, engine)
engine.dispose()
```

---

> What is the best practice and ideally could you link some open-source projects that implement it?

This could be done. But I am afraid this would lead to added complexity and the discussion diverging. Indeed, there would be a lot to get familiar with to reach the point about the decoupling of parsing and loading. That's because, as we have seen for `spaCy`, software projects who follow best practices tend to have part of the complexity of the challenges they tackle.

---

> I would be more than happy to make major changes:)

Actually, that wouldn't be major changes for decoupling parsing and loading. Indeed, this would just involve to go from:

```bash
# Current situation.
# Parsing and Loading articles is coupled.

# Parsing and Loading.
bbs_database add <database> <parser class> <data path>
```

to

```bash
# Adapted situation, for example.
# Parsing and Loading articles is decoupled.

# Parsing.
bbs_database parse <parser class> <raw data path>

# Loading.
bbs_database add <database> <parsed data path>
```

**Comment by jankrepl** [comment]
> The extensive example about decoupling parsing and loading is at the end of [#383 (review)](https://github.com/BlueBrain/Search/pull/383#pullrequestreview-687607172). It was part of the review of the PR introducing the `ArticleParser`. The specific part of the example I am thinking about is:
> 
> ```python
> # Parsing and Loading articles is decoupled.
> 
> # Parse articles.
> articles = parse_cord19(dirpath, nlp)
> 
> # Load articles in the database.
> engine = create_engine("mysql+pymysql://<user>:<password>@<host>:<port>/<database>")
> write_sql(articles, engine)
> engine.dispose()
> ```

So this is the desired solution in your opinion? I feel like it is identical to what is happening in the `add.py`. Or am I missing something?
https://github.com/BlueBrain/Search/blob/b8ff73b532140c2f2ad757693b1fe88c7df62abf/src/bluesearch/entrypoint/database/add.py#L84-L88


> > I would be more than happy to make major changes:)
> 
> Actually, that wouldn't be major changes for decoupling parsing and loading. Indeed, this would just involve to go from:
> 
> ```shell
> # Current situation.
> # Parsing and Loading articles is coupled.
> 
> # Parsing and Loading.
> bbs_database add <database> <parser class> <data path>
> ```
> 
> to
> 
> ```shell
> # Adapted situation, for example.
> # Parsing and Loading articles is decoupled.
> 
> # Parsing.
> bbs_database parse <parser class> <raw data path>
> 
> # Loading.
> bbs_database add <database> <parsed data path>
> ```

Interesting. However, what format is this raw data? Something as simple as a pickled instance of `Article`? Isn't this introducing an additionally layer of complexity?

**Comment by pafonta** [comment]
> > ```python
> > # Parsing and Loading articles is decoupled.
> > 
> > # Parse articles.
> > articles = parse_cord19(dirpath, nlp)
> > 
> > # Load articles in the database.
> > engine = create_engine("mysql+pymysql://<user>:<password>@<host>:<port>/<database>")
> > write_sql(articles, engine)
> > engine.dispose()
> > ```

> So this is the desired solution in your opinion?

This is a simplification of such desired solution. However, the general idea is there (decoupling parsing and loading).

> I feel like it is identical to what is happening in the `add.py`. Or am I missing something?
> https://github.com/BlueBrain/Search/blob/b8ff73b532140c2f2ad757693b1fe88c7df62abf/src/bluesearch/entrypoint/database/add.py#L84-L88

This isn't identical. In my suggestion, parsing (`parse_cord19`) and loading (`write_sql`) are a) both inside a function and b) could be called separately. In `add.py`, this is not the case for both a) and b).

---

> > ```shell
> > # Adapted situation, for example.
> > # Parsing and Loading articles is decoupled.
> > 
> > # Parsing.
> > bbs_database parse <parser class> <raw data path>
> > 
> > # Loading.
> > bbs_database add <database> <parsed data path>
> > ```
> 
> Interesting. However, what format is this raw data? Something as simple as a pickled instance of `Article`? Isn't this introducing an additionally layer of complexity?

A pickled instance of `Article` is a good start, yes.

**Comment by jankrepl** [comment]
> > > ```shell
> > > # Adapted situation, for example.
> > > # Parsing and Loading articles is decoupled.
> > > 
> > > # Parsing.
> > > bbs_database parse <parser class> <raw data path>
> > > 
> > > # Loading.
> > > bbs_database add <database> <parsed data path>
> > > ```
> > 
> > 
> > Interesting. However, what format is this raw data? Something as simple as a pickled instance of `Article`? Isn't this introducing an additionally layer of complexity?
> 
> A pickled instance of `Article` is a good start, yes.

Ok, one more question. Aren't we worried that a potential change inside of the source code of `Article` will make old pickles unloadable?

**Comment by pafonta** [comment]
> > > > ```shell
> > > > # Adapted situation, for example.
> > > > # Parsing and Loading articles is decoupled.
> > > > 
> > > > # Parsing.
> > > > bbs_database parse <parser class> <raw data path>
> > > > 
> > > > # Loading.
> > > > bbs_database add <database> <parsed data path>
> > > > ```
> > > 
> > > 
> > > Interesting. However, what format is this raw data? Something as simple as a pickled instance of `Article`? Isn't this introducing an additionally layer of complexity?
> > 
> > 
> > A pickled instance of `Article` is a good start, yes.
> 
> Ok, one more question. Aren't we worried that a potential change inside of the source code of `Article` will make old pickles unloadable?

This could indeed happen. However:

1 - Using a format different from Pickle will solve this. Using the format Pickle is just to start simple.

2 - Anyway, a change in the code of `Article` will break the whole pipeline, regardless of the decoupling, with the current implementation. The proposition I made in https://github.com/BlueBrain/Search/pull/383#pullrequestreview-687607172 solves this. Indeed, `write_sql` (see below) is generic regarding the loading of data. They could be articles or something else. But I suggest to not go for this topic in the current PR to stay focus on the first issue at hand: decoupling parsing and loading of articles.

```python
def write_sql(data: List[Data], engine: Engine, /) -> None:
```

File: src/bluesearch/entrypoint/database/parent.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/parent.py]: I am not really sure what the best practice is. Your suggestion looks reasonable, however, I will skip it for now. Feel free to copy paste your comment into a new issue and we can implement it in the future. Hope that is ok!

File: src/bluesearch/entrypoint/database/add.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/add.py]: Implemented in 89733d4bdfc57502e3e738dd7babce18209817f6

Mental note for all of us. Whenever we implement a new parser, we need to also modify the `parse` entrypoint.

File: tests/test_entrypoint/database/test_add.py
Comment (jankrepl) [reply tests/test_entrypoint/database/test_add.py]: Great point!
Fixed in f689cd9b828858d28d514412dadd594c196d1f23

File: tests/test_entrypoint/database/test_parent.py
Comment (jankrepl) [reply tests/test_entrypoint/database/test_parent.py:8]: Well, your suggestion cannot be used because `argparse` actually calls `sys.exit(0)` after it displays the help message. Which just kills the Python interpreter.
```python
# example.py
import argparse                                                                                                                                                                                       
                                                                                                                                                                                                      
parser = argparse.ArgumentParser()                                                                                                                                                                    
                                                                                                                                                                                                      
args = parser.parse_args(["-h"])                                                                                                                                                                      
print("Unreachable") 
```
```bash
python example.py
```
```bash
usage: example.py [-h]

optional arguments:
  -h, --help  show this help message and exit
```

Logically, `pytest` sees this as a failure. Even despite the fact that we exited with code 0.
```python
# test_example.py
import sys

def test():
    sys.exit(0)
    assert True

```
```bash
pytest test_example.py
```
```bash
...
FAILED test__example.py::test - SystemExit: 0
```

**Comment by jankrepl** [comment]
Ok, I made split the `add` command into `parse` and `add`. Could you guys review again? I updated the description.

File: src/bluesearch/entrypoint/database/add.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/add.py]:
```bash
python -X importtime -c "import bluesearch.database.article"  2> import.log
tuna import.log
```
Screen Shot 2021-08-13 at 10 22 28 AM

I guess the main problem is that `Python` will import all parent packages
`bluesearch.databse.article` -> `bluesearch.database` -> `bluesearch`. Unfortunately, `src/bluesearch/database/__init__.py` actually imports a lot of other stuff.

**Review by EmilieDel — APPROVED** [review]
Looks good to me 👍

File: src/bluesearch/entrypoint/database/add.py
Comment (EmilieDel) [reply src/bluesearch/entrypoint/database/add.py]: Really useful plot! Thanks @jankrepl !

Comment (EmilieDel) [reply src/bluesearch/entrypoint/database/add.py]: Let's create an issue for this and then resolve the conversation

File: src/bluesearch/entrypoint/database/parent.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/parent.py]: Created an issue #421

File: src/bluesearch/entrypoint/database/add.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/add.py]: Created an issue #420

File: src/bluesearch/entrypoint/database/parent.py
Comment (Stannislav) [inline src/bluesearch/entrypoint/database/parent.py:70]: Do we need this?

File: src/bluesearch/entrypoint/database/parse.py
Comment (Stannislav) [inline src/bluesearch/entrypoint/database/parse.py:53]: Just curious, what's the problem with types here?

**Review by Stannislav — APPROVED** [review]

File: src/bluesearch/entrypoint/database/parent.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/parent.py]: No, especially since we have the entrypoint inside of `setup.py`. Anyway, I deleted it here 2c8cacaa5654bdffa858f5820a2a2edcab7f39a1

File: src/bluesearch/entrypoint/database/parse.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/parse.py:53]:
```bash
src/bluesearch/entrypoint/database/parse.py:52: error: Incompatible types in assignment (expression has type "BinaryIO", variable has type "TextIO")  [assignment]
src/bluesearch/entrypoint/database/parse.py:53: error: Argument 2 to "dump" has incompatible type "TextIO"; expected "IO[bytes]"  [arg-type]
```

I did not really look into it too much:) If you know the recommended way around this feel free to share:)

**Comment by jankrepl** [comment]
Apparently #417 also did module renaming. The problematic one is `create_database.py` -> `database.py` since this PR actually introduces a subpackage called `database`. It is a mess and I open to find better namings/solutions. For now, I will just "revert" and go back to the `create_database.py`.

@Stannislav

File: src/bluesearch/entrypoint/database/parse.py
Comment (Stannislav) [reply src/bluesearch/entrypoint/database/parse.py:53]: I think renaming `f` to a different name (that was not used in the same scope before) should fix this.

I guess there was another variable called `f` somewhere before this block of code that was of type `BinaryIO`. `mypy` does not allow re-assigning variables with a different type.

**Comment by Stannislav** [comment]
> Apparently #417 also did module renaming. The problematic one is `create_database.py` -> `database.py` since this PR actually introduces a subpackage called `database`. It is a mess and I am open to finding better namings/solutions. For now, I will just "revert" and go back to the `create_database.py`.
> 
> @Stannislav

Oops, yes, sorry. Fine with me to go back to `create_database`.

File: src/bluesearch/entrypoint/database/parse.py
Comment (jankrepl) [reply src/bluesearch/entrypoint/database/parse.py]: You are right! I did not realize that the context manager actually lives on (is assigned to a variable) after being closed. Thank you! Fixed in 2a1801d6ba583ae15105b48e9e2b95dbd9c5bfd8

**Comment by FrancescoCasalegno** [comment]
## Experiment: Parse PubMed Central XML articles using CLI introduced in this PR

I tried to run the commands introduced with this PR to populate a database, using XML articles from PubMed Central.
To be precise, I only took articles belonging to journals with `"*neurosci*"` in the title.
If you are interested, here's the full code I used:
<details>
<summary>Code</summary>

```python
import os
from pathlib import Path
import time

import numpy as np
from tqdm import tqdm


PMC_PATH = Path("/raid/projects/bbs/pmc")
KEYWORD = "neurosci"
DATABASE = "neurosci.db"
PARSER = "PubmedXMLParser"


journals = [p for p in PMC_PATH.glob("*") if p.is_dir() and KEYWORD in p.name.lower()]

pcm_xml_files = [_ for journal in journals for _ in journal.rglob("*xml")]

os.system(f"bbs_database init {DATABASE}")

parse_times = []
add_times = []
tot_times = []
rm_times = []

for f_xml in tqdm(pcm_xml_files):
#    import pdb; pdb.set_trace()
    f_pkl = f_xml.stem + ".pkl"

    t0_ = time.time()
    try:
        t0 = time.time()
        os.system(f"""bbs_database parse {PARSER} "{f_xml}" "{f_pkl}" """)
        t1 = time.time()
        parse_times.append(t1-t0)
    except Exception as e:
        print(f"Failed PCM XML parsing on {f_xml}")
        print(e)
        continue

    try:
        t0 = time.time()
        os.system(f"""bbs_database add {DATABASE} "{f_pkl}" """)
        t1 = time.time()
        add_times.append(t1 - t0)
    except Exception as e:
        print(f"Failed adding article {f_xml}")
        print(e)
        continue

    t0 = time.time()
    os.remove(f_pkl)
    t1 = time.time()
    rm_times.append(t1 - t0)

    t1_ = time.time()
    tot_times.append(t1_ - t0_)

for stage, stage_name in [(parse_times, "Parsing"),
                          (add_times, "Adding"),
                          (rm_times, "Removing"),
                          (tot_times, "Total")]:
    print(f"{stage_name}: ")
    print(f"  - time [avg]    = {np.mean(stage):.4f}")
    print(f"  - time [median] = {np.median(stage):.4f}")
    print(f"  - time [min]    = {np.min(stage):.4f}")
    print(f"  - time [max]    = {np.max(stage):.4f}")
    print()
```
</details>

And these are the results I got on DGX1.
```bash
100%|████████████████████████████████████████████████████████████| 12083/12083 [1:26:34<00:00,  2.33it/s]
Parsing:
  - time [avg]    = 0.2058
  - time [median] = 0.2003
  - time [min]    = 0.1793
  - time [max]    = 0.3795

Adding:
  - time [avg]    = 0.2230
  - time [median] = 0.2168
  - time [min]    = 0.1955
  - time [max]    = 0.5332

Removing:
  - time [avg]    = 0.0007
  - time [median] = 0.0006
  - time [min]    = 0.0003
  - time [max]    = 0.0617

Total:
  - time [avg]    = 0.4295
  - time [median] = 0.4193
  - time [min]    = 0.3809
  - time [max]    = 0.7744
```

**Pull request merged**
